mole/services/ai.py

```diff
@@ -0,0 +1,411 @@
+from datetime import datetime
+from mole.common.help import InstructionHelper
+from mole.common.log import log
+from mole.common.task import BackgroundTask
+from mole.core.data import Path
+from mole.models.ai import (
+    AiVulnerabilityReport,
+    SeverityLevel,
+    tools,
+    VulnerabilityClass,
+    VulnerabilityReport,
+)
+from openai import OpenAI
+from openai.types.chat import (
+    ChatCompletionMessageParam,
+    ParsedChatCompletionMessage,
+    ParsedFunctionToolCall,
+)
+from typing import Any, Callable, Dict, Iterable, List, Optional, Tuple
+import binaryninja as bn
+import json
+import os
+import random
+import textwrap
+
+
+tag = "Mole.AI"
+
+
+class AiService(BackgroundTask):
+    """
+    This class implements a background task that analyzes paths using AI.
+    """
+
+    def __init__(
+        self,
+        bv: bn.BinaryView,
+        paths: List[Tuple[int, Path]],
+        analyzed_path: Callable[[int, AiVulnerabilityReport], None],
+        max_workers: Optional[int],
+        base_url: str,
+        api_key: str,
+        model: str,
+        max_turns: int,
+        max_completion_tokens: int,
+        initial_progress_text: str = "",
+        can_cancel: bool = False,
+    ) -> None:
+        """
+        This method initializes the AI service.
+        """
+        super().__init__(initial_progress_text, can_cancel)
+        self._bv = bv
+        self._paths = paths
+        self._analyzed_path = analyzed_path
+        self._max_workers = max_workers
+        self._base_url = base_url
+        self._api_key = api_key
+        self._model = model
+        self._max_turns = max_turns
+        self._max_completion_tokens = max_completion_tokens
+        self._tools = [tool.to_dict() for tool in tools.values()]
+        return
+
+    def _create_openai_client(self, custom_tag: str = tag) -> Optional[OpenAI]:
+        """
+        This method creates a new OpenAI client.
+        """
+        client = None
+        if self._base_url and self._api_key:
+            try:
+                client = OpenAI(base_url=self._base_url, api_key=self._api_key)
+            except Exception as e:
+                log.error(custom_tag, f"Failed to create OpenAI client: {str(e):s}")
+        return client
+
+    def _create_system_prompt(self) -> str:
+        vuln_class_lst = "\n".join(
+            f"   - `{vuln_class:s}`" for vuln_class in list(VulnerabilityClass)
+        )
+        sevr_level_lst = "\n".join(
+            f"   - `{sevr_level:s}`" for sevr_level in list(SeverityLevel)
+        )
+        tool_lst = "\n".join(f"- `{tool:s}`" for tool in tools.keys())
+        prompt = textwrap.dedent(
+            f"""You are a vulnerability research assistant specializing in static backward slicing of variables in Binary Ninja’s Medium Level Intermediate Language (MLIL) in Static Single Assignment (SSA) form. Your task is to determine whether a given slice corresponds to a real and exploitable vulnerability.
```

Comment:
Maybe we should give some suggestions about which code to get (HLIL, PseudoC, MLIL), i'm not even sure if it make sense for him to pull in different version of the same code besides maybe checking the stack variable size in the assembly